### PR TITLE
Add camera endpoint

### PR DESCRIPTION
**Pull request by azharmalik3**
@mosic

### CODE REVIEW

File: lib/evercam_media/snapshot/cam_client.ex [hunk 1]
```diff
@@ -27,6 +27,16 @@ defmodule EvercamMedia.Snapshot.CamClient do
     end
   end
 
+  def add_camera(args) do
+    #try do
+      models = Vendor.get_all_models(args[:vendor_exid])
+      models.vendor_models
```
Comment (azharmalik3) [inline lib/evercam_media/snapshot/cam_client.ex]: @mosic I am getting an error on this line. I want oo to check all models URL one by one.

File: .gitignore [hunk 2]
```diff
@@ -17,6 +17,7 @@ erl_crash.dump
 
 # Evercam specific files
 .env
+.idea
```
Comment (mosic) [inline .gitignore]: Make this change a separate commit.

File: web/models/vendor.ex [hunk 3]
```diff
@@ -1,10 +1,19 @@
 defmodule Vendor do
   use EvercamMedia.Web, :model
+  import Ecto.Query
+  alias EvercamMedia.Repo
 
   schema "vendors" do
-    has_many :vendor_models, VendorModel
+    has_many :vendor_model, VendorModel
```
Comment (mosic) [inline web/models/vendor.ex]: Revert this change.

File: web/models/vendor.ex [hunk 4]
```diff
@@ -1,10 +1,19 @@
...
 
     field :exid, :string
     field :name, :string
   end
+
+  def get_all_models(exid) do
```
Comment (mosic) [inline web/models/vendor.ex]: Rename this to `get_all`, like it's done on other models.

File: web/models/vendor.ex [hunk 5]
```diff
@@ -1,10 +1,19 @@
...
+    Vendor
+    |> where([vendor], vendor.exid == ^exid)
+    |> preload(:vendor_model)
+    |> Repo.first
```
Comment (mosic) [inline web/models/vendor.ex]: You're using `Repo.first` but this method is supposed to return all records. Use `Repo.all` instead.

File: web/router.ex [hunk 6]
```diff
@@ -33,6 +33,7 @@ defmodule EvercamMedia.Router do
     pipe_through :api
 
     post "/cameras/test", SnapshotController, :test
+    post "/cameras/add", SnapshotController, :add
```
Comment (mosic) [inline web/router.ex]: This should be in the `CameraController`, not the `SnapshotController`.

File: lib/evercam_media/snapshot/cam_client.ex (hunk 1 above)
Comment (mosic) [inline lib/evercam_media/snapshot/cam_client.ex]: @azharmalik3 See my other comments. For that line specifically, you need to use either `Enum.each` or a similar method in order to iterate over the collection. See http://elixir-lang.org/getting-started/enumerables-and-streams.html

File: web/models/vendor.ex (hunk 5 above)
Comment (azharmalik3) [inline web/models/vendor.ex]: The reason for using `Repo.first` is vendor record will be one but vendor_models will be more than one.

Comment (mosic) [inline web/models/vendor.ex]: Ah, ok then.

File: lib/evercam_media/snapshot/cam_client.ex [hunk 7]
```diff
@@ -27,9 +28,57 @@ defmodule EvercamMedia.Snapshot.CamClient do
...
 
+  def add_camera(args) do
+    try do
+      [username, password] = extract_auth_credentials(args)
+      models = Vendor.get_all_models(args[:vendor_exid])
+      Enum.each(models.vendor_models, fn model ->
+        snapshot_url = get_snapshot_url(args[:external_host], model.config["snapshots"]["jpg"], args[:external_http_port])
+        response = get_snapshot(args[:vendor_exid], snapshot_url, username, password)
+        Logger.debug(response)
+        #insert_camera(args, model.config["snapshots"]["jpg"])
```
Comment (azharmalik3) [inline lib/evercam_media/snapshot/cam_client.ex]: @mosic I want to exit from the loop when found working snapshot URL and then insert into the database. Could you please tell how to exit?

File: lib/evercam_media/snapshot/cam_client.ex [hunk 8]
```diff
@@ -27,9 +28,57 @@ defmodule EvercamMedia.Snapshot.CamClient do
...
+      end)
+      {:ok, Base.encode64("Success")}
+    rescue
+      error -> {:error, error}
+    end
+  end
 
   ## Private functions
 
+  defp get_snapshot(vendor, url, username, password) do
```
Comment (azharmalik3) [inline lib/evercam_media/snapshot/cam_client.ex]: @mosic this function always gave false response even got the snapshot.

File: lib/evercam_media/snapshot/cam_client.ex (hunk 7 above)
Comment (mosic) [inline lib/evercam_media/snapshot/cam_client.ex]: @azharmalik3 The proper way to do that is to use recursion instead: http://elixir-lang.org/getting-started/recursion.html

File: lib/evercam_media/snapshot/cam_client.ex (hunk 8 above)
Comment (mosic) [inline lib/evercam_media/snapshot/cam_client.ex]: @azharmalik3 That means you're getting an error somewhere? Use `IO.inspect` to determine the output of response before returning from a function to see where the problem is.

File: lib/evercam_media/snapshot/cam_client.ex (hunk 7 above)
Comment (azharmalik3) [inline lib/evercam_media/snapshot/cam_client.ex]: @mosic I wrote recursion functions and it works but I can't able to get the model by index. Try many ways to get first index value from `models.vendor_models` but it gave the error.

**Comment by azharmalik3** [comment]
@mosic could you please review branch again?

**Comment by mosic** [comment]
@azharmalik3 That's slightly better, but you're still doing it like you would in Ruby. Instead of using model count and keeping track of the index, take another look at how `head|tail` is used in Elixir guide: http://elixir-lang.org/getting-started/recursion.html

**Comment by azharmalik3** [comment]
@mosic please review my branch again?

File: lib/evercam_media/snapshot/cam_client.ex [hunk 9]
```diff
@@ -27,9 +28,84 @@ defmodule EvercamMedia.Snapshot.CamClient do
...
+    snapshot_url = get_snapshot_url(args[:external_host], head.config["snapshots"]["jpg"], args[:external_http_port])
+    response = get_snapshot(args[:vendor_exid], snapshot_url, username, password)
+    if response do
+      %{is_found: true, model: head}
+    else
+      find_model(tail, args)
+    end
+  end
+
+  defp get_snapshot(vendor, url, username, password) do
```
Comment (mosic) [inline lib/evercam_media/snapshot/cam_client.ex]: Why is this a duplicate `fetch_snapshot` function? Do you have a good reason for not reusing it?

File: lib/evercam_media/snapshot/cam_client.ex [hunk 10]
```diff
@@ -27,9 +28,84 @@ defmodule EvercamMedia.Snapshot.CamClient do
...
+    try do
+      response =
+        case vendor do
+          "evercam-capture" -> HTTPClient.get(:basic_auth_android, url, username, password)
+          "samsung" -> HTTPClient.get(:digest_auth, url, username, password)
+          "ubiquiti" -> HTTPClient.get(:cookie_auth, url, username, password)
+          _ -> HTTPClient.get(:basic_auth, url, username, password)
+        end
+        found_snapshor(response)
```
Comment (mosic) [inline lib/evercam_media/snapshot/cam_client.ex]: Typo.

File: lib/evercam_media/snapshot/cam_client.ex [hunk 11]
```diff
@@ -27,9 +28,84 @@ defmodule EvercamMedia.Snapshot.CamClient do
...
+      else
+        [username, password] = extract_auth_credentials(args)
+        models = Vendor.get_all_models(args[:vendor_exid])
+        total_models = Enum.count(models.vendor_models)
+        result = find_model(models.vendor_models, args)
+        if result.is_found do
+          camera = insert_camera(args, result.model)
+          {:ok, camera.exid}
+        else
+          {:not_found}
```
Comment (mosic) [inline lib/evercam_media/snapshot/cam_client.ex]: It would be better to use the `{:error, :not_found}` format here.

File: lib/evercam_media/snapshot/cam_client.ex [hunk 12]
```diff
@@ -27,9 +28,84 @@ defmodule EvercamMedia.Snapshot.CamClient do
     end
   end
 
+  def add_camera(args) do
+    try do
+      camera = Camera.by_exid(args[:camera_id])
+      unless is_nil(camera) do
+        {:duplicate}
```
Comment (mosic) [inline lib/evercam_media/snapshot/cam_client.ex]: Same as `:not_found` comment.

File: lib/evercam_media/snapshot/cam_client.ex [hunk 13]
```diff
@@ -65,6 +141,13 @@ defmodule EvercamMedia.Snapshot.CamClient do
     end
   end
 
+  defp get_snapshot_url(external_ur, snapshot_url, port) do
```
Comment (mosic) [inline lib/evercam_media/snapshot/cam_client.ex]: Typos in all occurrences of `external_url`.

File: web/router.ex [hunk 14]
```diff
@@ -43,6 +43,7 @@ defmodule EvercamMedia.Router do
     scope "/" do
       pipe_through :auth
 
+      post "/cameras/add", CameraController, :add
```
Comment (mosic) [inline web/router.ex]: Is this endpoints supposed to be different than `POST /v1/cameras` in `evercam-api`?

If not, this should be `POST /v1/cameras` and the route should be `:create`, not `:add`.
If yes, what are the differences?

File: web/models/vendor.ex [hunk 15]
```diff
@@ -1,10 +1,19 @@
...
   schema "vendors" do
     has_many :vendor_models, VendorModel
 
     field :exid, :string
     field :name, :string
   end
+
+  def get_all_models(exid) do
+    Vendor
+    |> where([vendor], vendor.exid == ^exid)
```
Comment (mosic) [inline web/models/vendor.ex]: You can simplify this as `|> where(exid: ^exid)`.

File: web/controllers/camera_controller.ex [hunk 16]
```diff
@@ -112,4 +122,84 @@ defmodule EvercamMedia.CameraController do
...
+
+  defp is_authorized(user) do
+    case user do
+      nil -> false
+      _ -> true
+    end
+  end
+
+  defp auto_generate_camera_id(camera_name) do
+    camera_name = camera_name |> String.replace(" ", "") |> String.replace("-", "") |> String.downcase |> String.slice(0..4)
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: Break this into multiple lines: https://github.com/niftyn8/elixir_style_guide

File: web/controllers/camera_controller.ex [hunk 17]
```diff
@@ -112,4 +122,84 @@ defmodule EvercamMedia.CameraController do
...
   end
+
+  ######################
+  ## Render functions ##
+  ######################
+
+  defp add_render(conn, 200, response) do
+    conn
+    |> put_status(200)
+    |> render("show.json", %{camera: response[:id] |> Camera.get_full, user: conn.assigns[:current_user]})
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: Don't use pipes inside of a struct, extract a variable out.

File: web/controllers/camera_controller.ex [hunk 18]
```diff
@@ -112,4 +122,84 @@ defmodule EvercamMedia.CameraController do
...
+    |> put_resp_header("access-control-allow-origin", "*")
+    |> json(response)
+  end
+
+  ##########################
+  ## Add Camera Functions ##
+  ##########################
+
+  defp add_camera(params, user) do
+    if is_authorized(user) do
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: Use `Permissions` module instead.

File: web/controllers/camera_controller.ex [hunk 19]
```diff
@@ -112,4 +122,84 @@ defmodule EvercamMedia.CameraController do
...
+  #####################
+
+  defp parse_add_camera_response({:ok, id}) do
+     [200, %{id: id}]
+  end
+
+  defp parse_add_camera_response(response) do
+    case response do
+      {:duplicate} ->
+        [409, %{message: "Camera exid already exist."}]
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: * `Camera ID`, not `Camera exid`
* `exists`, not `exist`

File: web/controllers/camera_controller.ex [hunk 20]
```diff
@@ -112,4 +122,84 @@ defmodule EvercamMedia.CameraController do
...
+      {:not_found} ->
+        [404, %{message: "Camera model not found."}]
+      _error ->
+        [500, %{message: "Sorry, we dropped the ball."}]
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: It would be useful to add log the error here so we know what went wrong with the request.

File: web/controllers/camera_controller.ex [hunk 21]
```diff
@@ -98,6 +99,15 @@ defmodule EvercamMedia.CameraController do
     end
   end
 
+  def add(conn, params) do
+    if is_nil(params["camera_id"]) do
+      params = Map.merge(%{"camera_id" => auto_generate_camera_id(params["camera_name"])}, params)
+    end
+    function = fn -> add_camera(params, conn.assigns[:current_user]) end
+    [code, response] = Util.exec_with_timeout(function, 15)
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: I know that you have looked at existing code as a reference so sorry about this complaint - but it's not necessary to use  forced timeout here. It would be actually better to not force the timeout anyway, so we can cycle through all vendor models before timing out.

File: lib/evercam_media/snapshot/cam_client.ex [hunk 22]
```diff
@@ -27,9 +28,84 @@ defmodule EvercamMedia.Snapshot.CamClient do
...
+    params = %{
+      exid: "#{camera_params[:camera_id]}",
+      name: camera_params[:camera_name],
+      owner_id: camera_params[:owner_id],
+      is_public: false,
+      is_online: true,
+      is_online_email_owner_notification: false,
+      model_id: model.id,
+      config: %{
+        external_host: camera_params[:external_host] |> String.to_atom,
```
Comment (mosic) [inline lib/evercam_media/snapshot/cam_client.ex]: Why convert to atoms? This should be a string.

File: web/controllers/camera_controller.ex [hunk 23]
```diff
@@ -112,4 +122,84 @@ defmodule EvercamMedia.CameraController do
...
+  defp auto_generate_camera_id(camera_name) do
+    camera_name = camera_name |> String.replace(" ", "") |> String.replace("-", "") |> String.downcase |> String.slice(0..4)
+    "#{camera_name}-#{Enum.take_random(?a..?z, 5)}"
+  end
+
+  ####################
+  ## Args functions ##
+  ####################
+
+  defp construct_args(params, user) do
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: Is there really a need for this function? You can just pass all `params` to `add_camera` function and parse them there.

File: web/controllers/camera_controller.ex [hunk 24]
```diff
@@ -112,4 +122,84 @@ defmodule EvercamMedia.CameraController do
...
+    |> json(response)
+  end
+
+  ##########################
+  ## Add Camera Functions ##
+  ##########################
+
+  defp add_camera(params, user) do
+    if is_authorized(user) do
+      res = construct_args(params, user)
```
Comment (mosic) [inline web/controllers/camera_controller.ex]: Move multiline pipeline to a new line: https://github.com/niftyn8/elixir_style_guide#source-code-layout

File: lib/evercam_media/snapshot/cam_client.ex [hunk 25]
```diff
@@ -27,9 +28,84 @@ defmodule EvercamMedia.Snapshot.CamClient do
...
+  defp find_model([], args) do
+    %{is_found: false}
+  end
+
+  defp find_model([head|tail], args) do
+    [username, password] = extract_auth_credentials(args)
+    snapshot_url = get_snapshot_url(args[:external_host], head.config["snapshots"]["jpg"], args[:external_http_port])
+    response = get_snapshot(args[:vendor_exid], snapshot_url, username, password)
+    if response do
+      %{is_found: true, model: head}
```
Comment (mosic) [inline lib/evercam_media/snapshot/cam_client.ex]: Change this to be `%{:ok, head}`.

File: lib/evercam_media/snapshot/cam_client.ex [hunk 26]
```diff
@@ -27,9 +28,84 @@ defmodule EvercamMedia.Snapshot.CamClient do
...
+      end
+    rescue
+      error -> {:error, error}
+    end
+  end
 
   ## Private functions
 
+  defp find_model([], args) do
+    %{is_found: false}
```
Comment (mosic) [inline lib/evercam_media/snapshot/cam_client.ex]: Change this to be `%{:error, :not_found}`.

File: lib/evercam_media/snapshot/cam_client.ex [hunk 27]
```diff
@@ -27,9 +28,84 @@ defmodule EvercamMedia.Snapshot.CamClient do
...
+    try do
+      camera = Camera.by_exid(args[:camera_id])
+      unless is_nil(camera) do
+        {:duplicate}
+      else
+        [username, password] = extract_auth_credentials(args)
+        models = Vendor.get_all_models(args[:vendor_exid])
+        total_models = Enum.count(models.vendor_models)
+        result = find_model(models.vendor_models, args)
+        if result.is_found do
```
Comment (mosic) [inline lib/evercam_media/snapshot/cam_client.ex]: Make the changes to `find_model`, then change this to use pattern matching:

``` elixir
case result do
  {:ok, model} ->
    ...
```

**Comment by mosic** [comment]
@azharmalik3 Good progress, you're almost getting there. See my comments.

File: web/router.ex (hunk 14 above)
Comment (azharmalik3) [inline web/router.ex]: No difference, both endpoints will create new camera.

Comment (marcoherbst) [inline web/router.ex]: This is the sort of silly thing that I might have suggested in the past in
order to comply with some standard that I hadn't properly thought about. Is
it?

If you don't see a need for 2x endpoints that do the same thing, then POST
/cameras seems the right one to me and it would be bad to have two that
duplicate each other and good to remove one of them.

Here's your reference:
https://docs.google.com/spreadsheets/d/1FakPt-hNWRyy0WiIrO1TFd2ay95uLBfx-Pfbdwbhedo/edit#gid=0

Comment (mosic) [inline web/router.ex]: @marcoherbst It might be. To me it makes the most sense for this endpoint to try the vendor model provided by the user, and only if the snapshot gives an error cycle through all other models and try to find a correct one. Right now this doesn't look at the vendor model parameter at all, just vendor. Is that intended?

Comment (marcoherbst) [inline web/router.ex]: Let's switch to voice, this doesn't make any sense to me. Milos I will call you.

File: lib/evercam_media/snapshot/cam_client.ex [hunk 28]
```diff
@@ -29,25 +29,25 @@ defmodule EvercamMedia.Snapshot.CamClient do
...
           {:ok, camera.exid}
         else
-          {:not_found}
+          {:error, :not_found}
         end
       end
-    rescue
-      error -> {:error, error}
-    end
+    #rescue
```
Comment (mosic) [inline lib/evercam_media/snapshot/cam_client.ex]: Always remove commented code.

File: lib/evercam_media/snapshot/cam_client.ex [hunk 29]
```diff
@@ -27,9 +28,87 @@ defmodule EvercamMedia.Snapshot.CamClient do
...
+        models = Vendor.get_all_models(args["vendor_id"])
+        result = find_model(models.vendor_models, args)
+        if result.is_found do
+          camera = insert_camera(args, result.model)
+          {:ok, camera.exid}
+        else
+          {:error, :not_found}
+        end
+      end
+    #rescue
```
Comment (mosic) [inline lib/evercam_media/snapshot/cam_client.ex]: Always remove commented code.

**Comment by azharmalik3** [comment]
This is just for exercise.So close now.
